### PR TITLE
ISM4A user management modules from ISIB-Group

### PR DESCRIPTION
**Pull request by ISIB-Group**
##### SUMMARY

Inspur server supports ansible management device  --user management
Basic management of the server based on restful interface

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

- is_add_user.py -  Create user.
- is_add_user_group.py - Create user group.
- is_del_user.py - Delete user 
- is_del_user_group.py - Delete user group
- is_edit_pass_user.py - Change user password
- is_edit_priv_user.py - Change user privilege
- is_edit_user_group.py - Set user group
- is_user_group_info.py -  Get user group information
- is_user_info.py - Get user information
##### ADDITIONAL INFORMATION

Circumstance instruction:
Ansible module is suitable for ansible version 2.2

Main steps:

Install suitable Ansible master
Install inspur_sdk 

Thes modules require the following to be installed on the Ansible server:

* Python 2.7 or later
* [Ansible](http://www.ansible.com) 2.2 or later
* [Inspur support] inspur_sdk  - Copy inspur_sdk to ansible plugins

```paste below
An example of static manifest for Inspur Server Manage is followed. The network functions is satisfied based on the assumed that Ansible moudle is available.

[root@localhost ansible]# ansible localhost -m is_user_info -a 'host=100.2.126.128 username=admin password=admin'
localhost | SUCCESS => {
    "Message": [
        {
            "User": [
                {
                    "Enable": "true", 
                    "Locked": "false", 
                    "Privilege": [
                        "KVM", 
                        "VMM"
                    ], 
                    "RoleId": "Administrator", 
                    "UserId": 1, 
                    "UserName": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
                }, 
                {
                    "Enable": "true", 
                    "Locked": "false", 
                    "Privilege": [
                        "KVM", 
                        "VMM"
                    ], 
                    "RoleId": "Administrator", 
                    "UserId": 2, 
                    "UserName": "root"
                }, 
                {
                    "Enable": "true", 
                    "Locked": "false", 
                    "Privilege": [], 
                    "RoleId": "Commonuser", 
                    "UserId": 3, 
                    "UserName": "yuwenjie"
                }, 
                {
                    "Enable": "true", 
                    "Locked": "false", 
                    "Privilege": [
                        "KVM"
                    ], 
                    "RoleId": "Operator", 
                    "UserId": 4, 
                    "UserName": "wbs"
                }
            ]
        }
    ], 
    "State": "Success", 
    "changed": false, 
    "proposed": {
        "host": "100.2.126.128", 
        "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
        "provider": null, 
        "subcommand": "getuser", 
        "username": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER"
    }
}


```

### CODE REVIEW

**Comment by felixfontein** [comment]
Hi, instead of adding a bunch of modules to community.general, how about creating your own collection? This would allow you to set up testing as you want, provide maintenance/support as you want, and release as you want, without being limited of the community releasing process.

**Comment by ISIB-Group** [comment]
The module we contribute to the community is based on THE SDK package, and the SDK package is not mentioned on Git in the module contribution. The download path provided by 20 results in the test failure. How do you deal with this situation?

**Comment by ISIB-Group** [comment]
@felixfontein The module we contribute to the community is based on THE SDK package, and the SDK package is not mentioned on Git in the module contribution. The download path provided by 20 results in the test failure. How do you deal with this situation?

**Comment by ISIB-Group** [comment]
@felixfontein The module we contribute to the community is based on THE SDK package. In the module contribution, the SDK package is not mentioned on Git, but the download path is provided, which results in the test failure. How do you deal with this situation?

**Comment by ISIB-Group** [comment]
@felixfontein hello,If I want to create my  own collection, how should I operate? Because it is the first time for community contribution, many processes are not very well understood. Could you please tell me?

**Comment by ISIB-Group** [comment]
@felixfontein We have developed 92 modules by ourselves about Inspur server supports ansible management device, and we want to contribute to the community, but due to the SDK's dependency on packages, the Shippable had part fails. Do you have any effective Suggestions in this regard? We look forward to your reply very much! thank you!

**Comment by felixfontein** [comment]
@ISIB-Group you can add the SDK here to make sure it is installed for the unit tests by ansible-test: https://github.com/ansible-collections/community.general/blob/master/tests/unit/requirements.txt

**Comment by felixfontein** [comment]
@ISIB-Group the fact that you created that may modules shows that you should really create your own collection. community.general (and community.network) are mostly for one-off modules, modules contributed by users for certain hardware/products, but not really for groups of modules developed by the vendor.

For documentation on creating a collection, see here: https://docs.ansible.com/ansible/devel/dev_guide/developing_collections.html#creating-collections

I think @gundalow also has an example somewhere on how to run tests with GitHub actions (so you don't need shippable).

**Comment by felixfontein** [comment]
Please also note that there are sanity check errors: https://app.shippable.com/github/ansible-collections/community.general/runs/2530/1/tests https://app.shippable.com/github/ansible-collections/community.general/runs/2530/2/tests https://app.shippable.com/github/ansible-collections/community.general/runs/2530/4/tests

The ansible-doc releated errors are because here in community.general, you need to create a symlink for every module from `plugins/modules/` to the correct module file.

**Comment by gundalow** [comment]
Hi, 
Thanks for the PR and great work. 
As @felixfontein said, I think this is well suited to having its own dedicated collection. We can help you get that setup. 

Are you part of Inspur, or do they know you are creating this? For new Collections we like there to be a few people that can help with it. We don't want to require lone individuals to have to carry all the work by themselves.

**Comment by ISIB-Group** [comment]
@gundalow Hello, we are the wave of a working group to work because of using ansible, so their developed some modules (there are 92) module is more, we'd like to have contributed to the ansible community originally, do you have any better suggestion, is to create your own collection or to the ansible contributed modules separately, if you can help set up, but the best, thank you very much

**Comment by gundalow** [comment]
Hi,
I'd create a new GitHub repo under https://github.com/Inspur soft called "ansible_collection"

**Comment by ISIB-Group** [comment]
@gundalow 好的，后续需要我们来做什么发信息给我们，非常谢谢

**Comment by ISIB-Group** [comment]
@gundalow Ok, what do you need us to do in the future? Send us a message. Thank you very much！

File: plugins/modules/remote_management/ism/is_add_user.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+# -*- coding:utf-8 -*-
+
+# Copyright(C) 2020 Inspur Inc. All Rights Reserved.
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+ANSIBLE_METADATA = {
```
Comment (felixfontein) [inline plugins/modules/remote_management/ism/is_add_user.py:11]: This should no longer be in modules.

File: plugins/modules/remote_management/ism/is_add_user.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+DOCUMENTATION = '''
+---
+module: is_add_user
+version_added: "2.9"
```
Comment (felixfontein) [inline plugins/modules/remote_management/ism/is_add_user.py:20]: This should be a collection version. The next release of community.general is 1.0.0.
```suggestion
version_added: "1.0.0"
```
For your own collection, you have to decide yourself what's the first released version is. I suggest starting with 0.1.0, and release 1.0.0 after some time when you are sure that no breaking changes will happen for some time. (Collection version numbers follow semantic versioning: https://semver.org/)

File: plugins/modules/remote_management/ism/is_add_user.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  vars:
+    ism:
+      host: "{{ ansible_ssh_host }}"
+      username: "{{ username }}"
+      password: "{{ password }}"
+
+  tasks:
+
+  - name: "add user"
+    is_add_user:
```
Comment (felixfontein) [inline plugins/modules/remote_management/ism/is_add_user.py:65]:
```suggestion
    community.general.is_add_user:
```
You should use FQCNs in examples (see https://github.com/ansible-collections/overview/#terminology).

For your own collection, the prefix would be different.

**Comment by ISIB-Group** [comment]
@felixfontein Hello, why did the test report syntax error after I added symlink for every module ? Do you know the reason

**Comment by felixfontein** [comment]
@ISIB-Group it looks like you did not add symlinks, but text files with the link target as their content.

**Comment by ISIB-Group** [comment]
@felixfontein Hello, I would like to ask you a question. I have defined parameter attributes in ISM file in Doc-fragments, but the error (" Argument 'password' in argument_spec defines type as 'str' but documentation doesn't define type") is still reported in the test. Do you know the reason?Thank you very much and look forward to your reply!

**Comment by ISIB-Group** [comment]
@felixfontein  All these questions are caused by this file ?
https://app.shippable.com/github/ansible-collections/community.general/runs/2594/2/tests

File: plugins/module_utils/ism.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+}
+ism_argument_spec = {
+    'provider': dict(type='dict', options=ism_provider_spec),
+}
+ism_top_spec = {
+    'host': dict(type='str'),
+    'username': dict(type='str'),
+    'password': dict(type='str', no_log=True),
+}
+ism_argument_spec.update(ism_top_spec)
```
Comment (felixfontein) [inline plugins/module_utils/ism.py:29]: Here you add `host`, `username` and `password` also to the top level. And for these you do not have entries in the docs fragment.

**Comment by ISIB-Group** [comment]
@felixfontein hello,the pull request checks have passed ,Shall we continue to wait for manual approval or there is another process? I am looking forward to your reply. Thank you!

File: plugins/modules/remote_management/ism/is_add_user.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        type: str
+        required: true
+    upass:
+        description:
+            - User password of new user.
+        type: str
+        required: true
+    role_id:
+        description:
+            - user role id of new user.
```
Comment (felixfontein) [inline plugins/modules/remote_management/ism/is_add_user.py:34]:
```suggestion
            - User role id of new user.
```
Descriptions should always be proper sentences, i.e. start with an upper-case letter (if it is not an identifier that is correctly written lower-case) and end with a period.

File: plugins/modules/remote_management/ism/is_add_user.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+      uname: "wbs"
+      upass: "admin"
+      role_id: "Administrator"
+      priv: "KVM,SOL"
+      provider: "{{ ism }}"
+'''
+
+RETURN = '''
+message:
+    description: messages returned after module execution
```
Comment (felixfontein) [inline plugins/modules/remote_management/ism/is_add_user.py:72]:
```suggestion
    description: Messages returned after module execution.
```
There are more (probably also in other files).

File: plugins/modules/remote_management/ism/is_del_user.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+version_added: "0.1.0"
+author:
+    - WangBaoshan (@ISIB-group)
+short_description: Delete user
+description:
+   - Delete user on Inspur server.
+options:
+    uname:
+        description:
+            - User name .
```
Comment (felixfontein) [inline plugins/modules/remote_management/ism/is_del_user.py:24]:
```suggestion
            - User name.
```

File: plugins/modules/remote_management/ism/is_user_info.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+  tasks:
+
+  - name: get user information
+    is_user_info:
+      provider: "{{ ism }}"
+'''
+
+RETURN = '''
+
```
Comment (felixfontein) [inline plugins/modules/remote_management/ism/is_user_info.py:46]: And where is the information about the user that is supposed to be returned?

**Comment by felixfontein** [comment]
> @felixfontein hello,the pull request checks have passed ,Shall we continue to wait for manual approval or there is another process? I am looking forward to your reply. Thank you!

Have you started the process of creating your own collection (resp. one for your company)? As @gundalow stated above, the plugins won't get merged to this collection.

**Pull request closed without merging**